tests/unit/data_classes/test_active_mq_event.py

```diff
@@ -6,33 +6,34 @@
 
 
 def test_active_mq_event():
-    event = ActiveMQEvent(load_event("activeMQEvent.json"))
+    raw_event = load_event("activeMQEvent.json")
+    parsed_event = ActiveMQEvent(raw_event)
 
-    assert event.event_source == "aws:amq"
-    assert event.event_source_arn is not None
-    assert len(list(event.messages)) == 3
+    assert parsed_event.event_source == raw_event["eventSource"]
+    assert parsed_event.event_source_arn == raw_event["eventSourceArn"]
+    assert len(list(parsed_event.messages)) == 3
 
-    message = event.message
+    message = parsed_event.message
     assert isinstance(message, ActiveMQMessage)
-    assert message.message_id is not None
-    assert message.message_type is not None
-    assert message.data is not None
-    assert message.decoded_data is not None
-    assert message.connection_id is not None
+    assert message.message_id == raw_event["messages"][0]["messageID"]
+    assert message.message_type == raw_event["messages"][0]["messageType"]
+    assert message.data == raw_event["messages"][0]["data"]
+    assert message.decoded_data == "ABC:AAAA"
+    assert message.connection_id == raw_event["messages"][0]["connectionId"]
     assert message.redelivered is False
-    assert message.timestamp is not None
-    assert message.broker_in_time is not None
-    assert message.broker_out_time is not None
-    assert message.properties["testKey"] == "testValue"
-    assert message.destination_physicalname is not None
+    assert message.timestamp == raw_event["messages"][0]["timestamp"]
+    assert message.broker_in_time == raw_event["messages"][0]["brokerInTime"]
+    assert message.broker_out_time == raw_event["messages"][0]["brokerOutTime"]
+    assert message.properties.get("testKey") == raw_event["messages"][0]["properties"]["testKey"]
+    assert message.destination_physicalname == raw_event["messages"][0]["destination"]["physicalname"]
     assert message.delivery_mode is None
     assert message.correlation_id is None
     assert message.reply_to is None
     assert message.get_type is None
     assert message.expiration is None
     assert message.priority is None
 
-    messages = list(event.messages)
+    messages = list(parsed_event.messages)
     message = messages[1]
```

Comment:
Actually there's a bug in ActiveMQ implementation for `message` which will **always** return the same message, since the generator is being redefined every time when accessing `.messages`.

---

More concise test to explain the bug

```python
class Messages:
    def __init__(self):
        self._messages = ["one", "two", "three"]
        ## NOTE: This defines a generator at instantiation
        # self.messages = (msg for msg in self._messages)

    ## NOTE: This defines a generator every time we access it
    ## which is fine for expected use in a loop (for msg in messages)
    ## but this doesn't work in the way `message` property is doing
    @property
    def messages(self):
        for msg in self._messages:
            yield msg

    @property
    def message(self):
        return next(self.messages)

data = Messages()
first_message = data.message
second_message = data.message

assert first_message != second_message
```

Reply:
Amazing catch @heitorlessa!

The same happens in the following data class:

https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/utilities/data_classes/kafka_event.py#L120

https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/utilities/data_classes/s3_event.py#L314

https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/utilities/data_classes/s3_event.py#L312

https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/utilities/data_classes/sns_event.py#L116

I'll try to change these files to see if it is a potential breaking change or not.

Reply:
Hi @heitorlessa!! Check out this commit - [`d8b2739` (#2506)](https://github.com/aws-powertools/powertools-lambda-python/pull/2506/commits/d8b273994543f75dbb6f52ab51383b263be9663d)

I addressed this feedback and fixed this property in the `ActiveMQ` and `Kafka` event source classes. For S3, SNS and SES don't need to change because it is always just one record per Lambda trigger.

Thanks for giving this feedback, it was very helpful to learn new things.